invoices: Add `paid_timestamp` field.

ZmnSCPxj opened the pull request:
Fixes: #615

cdecker commented on wallet/invoices.c:
```diff
@@ -61,8 +61,10 @@ static bool wallet_stmt2invoice(sqlite3_stmt *stmt, struct invoice *inv)
-	if (inv->state == PAID)
+	if (inv->state == PAID) {
 		inv->msatoshi_received = sqlite3_column_int64(stmt, 8);
+		inv->paid_timestamp = sqlite3_column_int64(stmt, 9);
```
Sort of uses implicit NULL -> 0 conversion, should be fine though.

ZmnSCPxj replied:
`paid_timestamp` should get set at the point that `state` becomes `PAID`, xref `msatoshi_received`, which is similar.

We are actually slightly inconsistent here, since `pay_index` is `NULL` on-db if `state != PAID` too, and yet we always load it.

cdecker commented on wallet/wallet.h:
```diff
@@ -363,6 +363,8 @@ struct invoice {
 	/* Set if state == PAID */
 	u64 msatoshi_received;
+	/* Set if state == PAID */
+	u64 paid_timestamp;
```
Can't find it being set anywhere before `state == PAID`, should probably be initialized to `0` [here](https://github.com/ZmnSCPxj/lightning/blob/51ba88213fe2cd67ee20dcb6b1e49b9d7412d878/wallet/invoices.c#L184) just to mimic loading from DB as well.

ZmnSCPxj replied:
`msatoshi_received` is similar also (i.e. not set anywhere before `state == PAID`).  indeed `pay_index` and `msatoshi_received` are not handled similarly (`pay_index` gets set to 0 always, `msatoshi_received` is only set if `state == PAID`), so we should reconcile those two first.

ZmnSCPxj replied:
In theory, not loading the fields should help valgrind detect cases where the program erroneously reads those fields when `state != PAID`.  So I suppose I can understand some rationale for not loading a field that has no valid value at  this point anyway.

rustyrussell replied:
Yes, I prefer non-initialization for exactly this reason, even when loading from db.

I pondered adding an "uninitialized(x)" macro which would memset to 0 (or 0x42424242...) iff not running under valgrind (which is *usually* less disasterous than leaking some random value).

ZmnSCPxj replied:
Okay, I will add a commit to *not* set `pay_index` if not PAID instead.

rustyrussell replied:
Hmm, I'm going to break this with the JSON API patches :(

But you'll also need to add the field to doc/lightning-listinvoices.7.txt after that...

ZmnSCPxj replied:
Maybe just cherry-pick this particular commit and add it to your JSON API pull request?  Probably after removing the unnecessary "CREATE INDEX invoices_paid_timestamp".  Then I'll just put the "not set `pay_index` if not PAID" in a separate pull request.

ZmnSCPxj commented:
Rebased on master.

rustyrussell commented on wallet/db.c:
```diff
@@ -177,6 +177,14 @@ char *dbmigrations[] = {
+    "ALTER TABLE invoices ADD paid_timestamp INTEGER;",
+    "CREATE INDEX invoices_paid_timestamp"
+    "    ON invoices(paid_timestamp);",
+    "UPDATE invoices"
```
Why create this index?

ZmnSCPxj replied:
I was thinking it would be useful for sorting according to timestamp, but I realize now that sorting according to `pay_index` would be the same anyway.  I will remove this.

rustyrussell commented on wallet/db.c:
```diff
@@ -177,6 +177,14 @@ char *dbmigrations[] = {
+    "    ON invoices(paid_timestamp);",
+    "UPDATE invoices"
+    "   SET paid_timestamp = strftime('%s', 'now')"
+    " WHERE state = 1;",
```
Hmm, this is a hard one.  We could leave it as zero, to clearly flag that it was pre-db-update?

Officially, we don't support db updates yet, but OTOH this is all good practice...

ZmnSCPxj replied:
hmm indeed....

ZmnSCPxj commented:
Removed unnecessary index on `paid_timestamp` (ordering by `pay_index` is better) and also added a separate commit to not initialize `pay_index` if not PAID.

cdecker commented:
I'll try to rebase @rustyrussell's WIP JSON API fixes on top of this, and if it works I'll merge this and update that PR

cdecker commented:
Ok, rebase of #623 worked out nicely, so I'll merge this now and push the rebase on top of `master` for that one.
